### PR TITLE
Replace "hard-errors" with NumbaErrors.

### PR DESCRIPTION
**Pull request by stuartarchibald**
This patch aims to replace the remaining "hard-errors" that may have been reachable from code running type-inference with appropriate `NumbaError` instances. This is to make the code base internally compliant with the permanent change to "new_style" error handling.

### CODE REVIEW

File: numba/core/compiler_machinery.py [hunk 1]
```diff
@@ -355,7 +355,7 @@ def run(self, state):
                 if isinstance(pass_inst, CompilerPass):
                     self._runPass(idx, pass_inst, state)
                 else:
-                    raise BaseException("Legacy pass in use")
+                    raise RuntimeError("Legacy pass in use")
```
Comment (sklam) [inline numba/core/compiler_machinery.py:358]: Suggest to undo this change.
The use of `BaseException` is to make it not catchable by `except Exception`.

Comment (stuartarchibald) [reply numba/core/compiler_machinery.py]: Fixed in 6e47aa2

**Review by sklam — APPROVED** [review]
Thanks for this very tedious effort in identifying all problematic `raise`

**Comment by stuartarchibald** [comment]
> Thanks for this very tedious effort in identifying all problematic `raise`

No problem, thanks for the review. Your code that did the tracing helped a load and improved my confidence about having found most of them, thanks for writing it!
